rust-codegen/codegen/src/parser.rs

```diff
@@ -0,0 +1,450 @@
+use std::str;
+use std::path::{Path, PathBuf};
+
+use types::{Enumerator, Field, FieldType, FileDescriptor, Frequency, Message, snake_case};
+use nom::{digit, hex_digit, multispace};
+
+fn is_word(b: u8) -> bool {
+    match b {
+        b'a'...b'z' | b'A'...b'Z' | b'0'...b'9' | b'_' | b'.' => true,
+        _ => false,
+    }
+}
+
+named!(
+    word<String>,
+    map_res!(take_while!(is_word), |b: &[u8]| String::from_utf8(
+        b.to_vec()
+    ))
+);
+named!(
+    word_ref<&str>,
+    map_res!(take_while!(is_word), str::from_utf8)
+);
+
+named!(
+    hex_integer<i32>,
+    do_parse!(
+        tag!("0x")
+            >> num:
+        map_res!(
+            map_res!(hex_digit, str::from_utf8),
+            |s| i32::from_str_radix(s, 16)
+        ) >> (num)
+    )
+);
+
+named!(
+    integer<i32>,
+    map_res!(map_res!(digit, str::from_utf8), str::FromStr::from_str)
+);
+
+named!(
+    comment<()>,
+    do_parse!(tag!("//") >> take_until_and_consume!("\n") >> ())
+);
+named!(
+    block_comment<()>,
+    do_parse!(tag!("/*") >> take_until_and_consume!("*/") >> ())
+);
+
+/// word break: multispace or comment
+named!(
+    br<()>,
+    alt!(map!(multispace, |_| ()) | comment | block_comment)
+);
+
+named!(
+    package<String>,
+    do_parse!(
+        tag!("package") >> many1!(br) >> package: word >> many0!(br) >> tag!(";") >> (package)
+    )
+);
+
+named!(
+    num_range<Vec<i32>>,
+    do_parse!(
+        from_: integer >> many1!(br) >> tag!("to") >> many1!(br) >> to_: integer
+            >> ((from_..(to_ + 1)).collect())
+    )
+);
+
+named!(
+    frequency<Frequency>,
+    alt!(tag!("optional") => { |_| Frequency::Optional } |
+         tag!("repeated") => { |_| Frequency::Repeated } |
+         tag!("required") => { |_| Frequency::Required } )
```

Comment:
These aren't part of the Kiwi format. Please remove code unrelated to Kiwi from this file.